crates/dbs-legacy-devices/src/rtc_pl031.rs

```diff
@@ -0,0 +1,301 @@
+        (ts / dbs_utils::time::NANOS_PER_SECOND as i128) as u32
+    }
+
+    fn handle_write(&mut self, offset: u64, val: u32) -> Result<()> {
+        match offset {
+            RTCMR => {
+                // The MR register is used for implementing the RTC alarm. A real time clock alarm is
+                // a feature that can be used to allow a computer to 'wake up' after shut down to execute
+                // tasks every day or on a certain day. It can sometimes be found in the 'Power Management'
+                // section of a motherboard's BIOS setup.
+                self.match_value = val;
+                self.metrics.missed_write_count.inc();
+            }
+            RTCLR => {
+                self.load = val;
+                self.previous_now = Instant::now();
+                // If the unwrap fails, then the internal value of the clock has been corrupted and
+                // we want to terminate the execution of the process.
+                self.tick_offset = dbs_utils::time::seconds_to_nanoseconds(i64::from(val))
+                    .expect("Time conversion overflow");
+            }
+            RTCIMSC => {
+                self.imsc = val & 1;
+                self.trigger_interrupt()?;
+            }
+            RTCICR => {
+                // As per above mentioned doc, the interrupt is cleared by writing any data value to
+                // the Interrupt Clear Register.
+                self.ris = 0;
+                self.trigger_interrupt()?;
+            }
+            RTCCR => (), // ignore attempts to turn off the timer.
+            o => {
+                return Err(Error::BadWriteOffset(o));
+            }
+        }
+        Ok(())
+    }
+    /// Provides a reference to the interrupt event fd.
+    pub fn interrupt_evt(&self) -> &EventFd {
+        &self.interrupt_evt
+    }
+}
+
+impl DeviceIoMut for RTCDevice {
+    fn read(&mut self, _base: IoAddress, offset: IoAddress, data: &mut [u8]) {
+        let v;
+        let mut read_ok = true;
+        let offset = offset.raw_value();
+
+        if offset < AMBA_ID_HIGH && offset >= AMBA_ID_LOW {
+            let index = ((offset - AMBA_ID_LOW) >> 2) as usize;
+            v = u32::from(PL031_ID[index]);
+        } else {
+            v = match offset {
+                RTCDR => self.get_time(),
+                RTCMR => {
+                    // Even though we are not implementing RTC alarm we return the last value
+                    self.metrics.missed_read_count.inc();
+                    self.match_value
+                }
+                RTCLR => self.load,
+                RTCCR => 1, // RTC is always enabled.
+                RTCIMSC => self.imsc,
+                RTCRIS => self.ris,
+                RTCMIS => self.ris & self.imsc,
+                _ => {
+                    read_ok = false;
+                    0
+                }
+            };
+        }
+        if read_ok && data.len() <= 4 {
```

Comment:
Why is it "<= 4" instead of "== 4"?

Reply:
This should be "== 4" and we will use the vm_superio crate to refactor this file.